lib/ack-tracker/batched_ack_tracker.c

```diff
@@ -0,0 +1,261 @@
+
+  return &(self->pending_ack_record->super.bookmark);
+}
+
+static void
+_assign_pending_ack_record_to_msg(BatchedAckTracker *self, LogMessage *msg)
+{
+  LogSource *source = self->super.source;
+  log_pipe_ref((LogPipe *) source);
+  self->pending_ack_record->super.bookmark.persist_state = source->super.cfg->state;
+  self->pending_ack_record->super.tracker = &self->super;
+  msg->ack_record = (AckRecord *) self->pending_ack_record;
+}
+
+static void
+_track_msg(AckTracker *s, LogMessage *msg)
+{
+  BatchedAckTracker *self = (BatchedAckTracker *)s;
+  _assign_pending_ack_record_to_msg(self, msg);
+  self->pending_ack_record = NULL;
+}
+
+static GList *
+_append_ack_record_to_batch(BatchedAckTracker *self, AckRecord *ack_record)
+{
+  GList *full_batch = NULL;
+  g_mutex_lock(&self->acked_records_lock);
+  {
+    self->acked_records = g_list_prepend(self->acked_records, ack_record);
+    ++self->acked_records_num;
+    if (self->acked_records_num == self->batch_size)
+      {
+        full_batch = self->acked_records;
+        self->acked_records = NULL;
+        self->acked_records_num = 0;
+      }
+  }
+  g_mutex_unlock(&self->acked_records_lock);
+
+  return full_batch;
+}
+
+static void
+_stop_batch_timer(BatchedAckTracker *self)
+{
+  if (iv_timer_registered(&self->batch_timer))
+    iv_timer_unregister(&self->batch_timer);
+}
+
+static void
+_start_batch_timer(BatchedAckTracker *self)
+{
+  if (self->timeout <= 0)
+    return;
+
+  iv_validate_now();
+  self->batch_timer.expires = iv_now;
+  timespec_add_msec(&self->batch_timer.expires, self->timeout);
+
+  iv_timer_register(&self->batch_timer);
+}
+
+static void
+_batch_timeout(gpointer data)
+{
+  BatchedAckTracker *self = (BatchedAckTracker *) data;
+  GList *batch = NULL;
+  g_mutex_lock(&self->acked_records_lock);
+  {
+    batch = self->acked_records;
+    self->acked_records = NULL;
+    self->acked_records_num = 0;
+  }
+  g_mutex_unlock(&self->acked_records_lock);
+  if (batch)
+    {
+      _ack_batch(self, batch);
+    }
+}
+
+static void
+_init_watches(BatchedAckTracker *self)
+{
+  IV_TIMER_INIT(&self->batch_timer);
+  self->batch_timer.cookie = self;
+  self->batch_timer.handler = _batch_timeout;
+}
+
+static void
+_manage_msg_ack(AckTracker *s, LogMessage *msg, AckType ack_type)
+{
+  BatchedAckTracker *self = (BatchedAckTracker *) s;
+  log_source_flow_control_adjust(self->super.source, 1);
+
+  if (ack_type == AT_SUSPENDED)
+    log_source_flow_control_suspend(self->super.source);
+
+  if (ack_type != AT_ABORTED)
+    {
+      if (!iv_timer_registered(&self->batch_timer))
+        _start_batch_timer(self);
+      GList *full_batch = _append_ack_record_to_batch(self, msg->ack_record);
+      if (full_batch)
+        _ack_batch(self, full_batch);
+    }
+  else
+    {
+      _ack_record_free(msg->ack_record);
+    }
+
+  log_msg_unref(msg);
+  log_pipe_unref((LogPipe *)self->super.source);
+}
+
+static void
+_disable_bookmark_saving(AckTracker *self)
+{
+  // TODO -> ??
```

Comment:
What will be the future of this function? Is it needed?

Reply:
I think, it is - I need some time to finalize this.

Reply:
This is used only by `FileReader`  (`file_reader_stop_follow_file` - which is used only by `WildcardSource`), so no, it won't be supported by `BatchedAckTracker`.